DOC Add minor release policy (#308)
silverstripe/developer-docs: en/06_Project_Governance/06_Minor_release_policy.md

Note that the release cadence and pre-release time frame are indicative only. We may alter those in response to external constraints.

Our minor release policy is more flexible because minor release upgrades are more straight forward to manage than major release upgrades. 

## Upgrading to a new minor release

Silverstripe CMS follows [semantic versioning](https://semver.org/). Silverstripe CMS minor releases deliver new features and new APIs in a backward compatible way.

## Minor release cadence

Silverstripe CMS aims to ship a new minor releases every 6 months for the major release line in active development. Those minor releases are targeted for April and October each year.

### Pre-release

Silverstripe CMS does not usually tag alpha releases for minor releases.

Approximately 6 weeks prior to the anticipated stable minor release, a beta minor release is published. Once a beta release is tagged, any new feature or API that didn't make it to the beta should be targeted to the follow up minor release. This allows the CMS development team to perform a regression test on the beta release with confidence that no additional regressions will be introduced before the stable release.

> 
> 

Reviewer: ```suggestion

```

Author: Done.

Approximately 2 weeks prior to the anticipated stable minor release, a release candidate is tagged. Once a release candidate is tagged, only critical impact bug fixes can be added to the release. The release candidate is sent to an external auditor for a security review. Any security patches which will be included in the stable release are also sent to the auditor.

### Stable release

New minor releases of Silverstripe CMS are tagged around April and October each year.

## Minor release support timeline

The minor release support timeline follows similar phases to those of a major release, but are more condensed.

### Bug and security fixes

A Silverstripe CMS minor release line enters the *bug and security fixes* phase once it is tagged *stable*.

A minor release in the *bug and security fixes* phase receives:

> - regular patches for bugs at all impact levels
> - security patches for all vulnerabilities.

Reviewer: ```suggestion
- bugfixes and enhancements that do not change existing APIs
- security patches for vulnerabilities at all impact levels.
```

Author: Applied the suggestion mostly as-is, but removed the bit about "enhancements". I could imagine some pretty substantial enhancement that wouldn't change APIs that we would never ship in a patch.

It does **not** receive:

> - new feature
> - new APIs.

Reviewer: ```suggestion
- new features
- new APIs
```

Author: That follows our current recommended style guide. https://silverstripe.atlassian.net/wiki/spaces/MAR/pages/506364107/Silverstripe+Writing+Style+Guide#Lead-in-to-bulleted-lists

Reviewer 2: I agree the fullstop should remain.
The change to plurality still needs to be made.

Author: Fixed.

### Security fixes only

Once the follow up minor release is tagged, the minor release that was in the *bug and security fixes* phase transition to a *security fixes only* phase.

In the *security fixes only* phase, a Silverstripe CMS minor release line only receives patches for high impact security issues defined has having a CVSS of 7.0 or above.

### End-of-life

Six months after entering the *security fixes only* phase, the minor release goes *end-of-life*. No more security patches will be shipped for that minor release.

### The last minor release in a major release line

Shortly before tagging a new major release, one last minor release will be tagged for the major release line in active development.

The last minor release in a major release line will have an extended support phases in line with our major release policy:
- the last minor release stays in the *bug and security fixes* phase for one year
- then it enters the *security fixes only* phase for another year
- then the minor release goes end of life along with its corresponding major release line.

## Security patch windows

> Silverstripe CMS aims to deliver security patches at pre-defined intervals to minimise upgrade woes for Silverstripe CMS projects.

Reviewer: ```suggestion
Silverstripe CMS aims to deliver security patches at pre-defined intervals.
```

Reviewer: You don't deliver security patches to make it easier to upgrade, you deliver them to deliver security patches :-)

Reviewer 2: I read this as meaning "we have a specific cadence of releasing patches at pre-defined intervals instead of just releasing these ad-hoc to minimise upgrade woes for Silverstripe CMS projects."

Reviewer: OK maybe change "to minimise upgrade woes for Silverstripe CMS projects" to "to make it easier to plan upgrading Silverstripe CMS projects so they can receive the security patches"

Author: Went with:
> Silverstripe CMS tags patch releases of individual modules whenever relevant bug fixes or other very low risk changes are merged into a supported patch release branch.

When a vulnerability is [confidentially reported](/contributing/issues_and_bugs/#reporting-security-issues) to the Silverstripe CMS development team and we are confident that the vulnerability is not actively being exploited in the wild, we will wait until the next available security patch window to disclose it and release a patch.

Security patch windows are scheduled to line up with our expected minor release. Our security patch windows are in:
- January
- April
- July
- October.

Any security fixes that are shipped at the same time as a new minor release will also be patched and tagged on the minor release that is entering the "bug and security fixes" phase.

## Patch releases

> Silverstripe CMS tags patch releases of individual modules whenever relevant bug fixes or other very low risk changes are merged into the codebase.

Reviewer: ```suggestion
Silverstripe CMS tags patch releases of individual modules whenever relevant bug fixes or other changes are merged into a currently supported minor branch of the codebase.
```

Author: There was multiple and contradictory threads on this specific line. I think the current version of the line encapsulate everyone's concerns.

---
title: Minor Release Policy
summary: Outline the minor release lifecycle and support commitment for Silverstripe CMS
icon: code-branch
---

# Minor Release Policy

## Scope of the policy

This policy applies to all [Silverstripe CMS commercially supported modules](/project_governance/supported_modules/).

Community modules are not covered by this policy. Modules in the `silverstripe` github organisation that are not commercially supported are updated on a best effort basis.

> ## General approach to major releases

Reviewer 2: ```suggestion
## General approach to minor releases
```
lol

Author: Should be fixed now.

> ## General approach to major releases

Reviewer 2: What is the point of this section?

In the major release policy it outlines... well, the general approach to releases. It says what the support levels are, and how long they last - and explains what is expected to be in those releases.

This section should either be renamed, or should reflect that same sort of information.

Author: Content as been moved to the top. Header has been removed.

> This policy complement the [Silverstripe CMS Major release policy](major_release_policy) by providing medium term guidance on when new minor releases of Silverstripe CMS are shipped.

Reviewer 2: ```suggestion
This policy complements the [Silverstripe CMS Major release policy](major_release_policy) by providing guidance on when new minor releases of Silverstripe CMS are shipped.
```
I don't think "medium term" adds anything here and is a little confusing as to what it means.

Author: Done

> This policy complement the [Silverstripe CMS Major release policy](major_release_policy) by providing medium term guidance on when new minor releases of Silverstripe CMS are shipped.

Reviewer 2: This doesn't sound like it's about the "general approach to minor releases" - it sounds like it belongs in the scope section, or perhaps directly under the title.

Author: You are right this entire section makes more sense as a introduction to the policy

Note that the release cadence and pre-release time frame are indicative only. We may alter those in response to external constraints.

Our minor release policy is more flexible because minor release upgrades are more straight forward to manage than major release upgrades. 

> ## Upgrading to a new major release

Reviewer 2: ```suggestion
## Upgrading to a new minor release
```

Author: Fixed.

Silverstripe CMS follows [semantic versioning](https://semver.org/). Silverstripe CMS minor releases deliver new features and new APIs in a backward compatible way.

> ## Minor release cadence

Reviewer 2: Why "cadence" instead of "lifecycle" (the major release policy says "lifecycle")?

Author: The lifetime of a minor is much shorter and simpler. Cadence seem to be more appropriate here, but not fuss either way.

Silverstripe CMS aims to ship a new minor releases every 6 months for the major release line in active development. Those minor releases are targeted for April and October each year.

### Pre-release

> Silverstripe CMS does not usually tag alpha releases for minor releases. Changes intended to be shipped in the next minor release are tracked in *minor release branches* for each module. Those branches are normally named to an integer value. For example, the next minor release branch of `silverstripe/installer` 5 is called `5` and can be installed with `composer create-project silverstripe/installer:5.x-dev mysite`.

Reviewer 2: We don't need to explain our branching strategy or release process here.
```suggestion
Silverstripe CMS does not usually tag alpha releases for minor releases.
```

Author: Suggestion applied.

> Approximatively 6 weeks prior to the anticipated stable minor release, a beta minor release is published. Once a beta release is tagged, any new feature or API that didn't make it to the beta should be targeted to the follow up minor release. This allows the CMS development team to perform a regression test on the minor release. 

Reviewer 2: ```suggestion
Approximately 6 weeks prior to the anticipated stable minor release, a beta minor release is published. Once a beta release is tagged, any new feature or API that didn't make it to the beta should be targeted to the follow up minor release. This allows the CMS development team to perform a regression test on the beta release with confidence that no additional regressions will be introduced before the stable release.
```

Author: Suggestion applied.

> The next minor release branch of each module is forked to a *patch release branch*. Patch release branches normally follow the pattern `MAJOR.MINOR`. For example, the patch release branch for `silverstripe/installer` 5.1 is called `5.1` and can be installed with `composer create-project silverstripe/installer:5.1.x-dev mysite`.

Reviewer 2: We don't need to explain our branching strategy or release process here.
```suggestion
```

Author: Suggestion applied.

> Approximatively 2 weeks prior to the anticipated stable minor release, a release candidate is tagged. The release candidate is sent to an external auditor for a security review. The auditor normally received confidential security patches as well for review.

Reviewer 2: ```suggestion
Approximately 2 weeks prior to the anticipated stable minor release, a release candidate is tagged. Once a release candidate is tagged, only critical impact bug fixes can be added to the release. The release candidate is sent to an external auditor for a security review. Any security patches which will be included in the stable release are also sent to the auditor.
```
I _think_ it's critical ones - but maybe high? Can't remember.

Author: Suggestion applied

### Stable release

> Once the team is ready, stable versions of all supported modules are tagged. A complete changelog is posted to the Silverstripe CMS documentation site along with any relevant upgrade notes.

Reviewer 2: No need to say "When we're ready" etc - that's all part of the release process, not part of the policy.
```suggestion
New minor releases of Silverstripe CMS are tagged around April and October each year.
```
If we're mentioning a changelog here we should also mention it in the major release policy.

Author: Suggestion applied.

> ## Minor release support timeline

Reviewer 2: Where's the "active development" phase?

Author: In the context of minor that would be nonsensical. We don't add new feature to minor, so they can't be in active development post-stable.

Reviewer 2: They're in active development before they're ever tagged pre-stable.

Author: That's not how we use "active development" for majors. e.g.

- CMS 5 is in active development and stable right now.
- CMS 5.0 is not in active development, but is stable.

We've historically only applied "active development" to majors.

> Minor release support timeline follow similar phases to major release. But those phases are far more condensed.

Reviewer 2: "those are farm more condensed" is ambiguous - is it the major or the minor timeline that is more condensed?
```suggestion
The minor release support timeline follows similar phases to those of a major release, but are more condensed.
```

Author: Suggestion applied.

### Bug and security fixes

A Silverstripe CMS minor release line enters the *bug and security fixes* phase once it is tagged *stable*.

A minor release in the *bug and security fixes* phase receives:
- regular patches for bugs at all impact levels
- security patches for all vulnerabilities.

It does **not** receive:
- new feature
- new APIs.

A minor release line stays in the *bug and security fixes* phase until a follow up minor release is tagged.

### Security fixes only

> Once the follow up minor release is tagged, the minor release that was in the *bug and security fixes* phase transition to a *security fixes only* phase.

Reviewer 2: The wording from the major release policy is clearer.
```suggestion
At the end of the *bug and security fixes* phase, a Silverstripe CMS minor release line transitions to the *security fixes only* phase.
```

Author: The minor's transition is triggered by the release of another minor.

The major's transition is triggered by a calendar date. This nuance seems important here.

Reviewer 2: That nuance is already covered by saying _when_ the "bug and security fixes" ends in the section about "bug and security fixes".

This next phase starts _when the last one ends_.

Author: Applied suggestion

In the *security fixes only* phase, a Silverstripe CMS minor release line only receives patches for high impact security issues defined has having a CVSS of 7.0 or above.

> Upon entering the *security fixes only* phase, the end-of-life date for that minor release is scheduled for six months later.

Reviewer 2: This is mentioned directly below. Either don't mention it here at all, or mention it in a way that centers around _this phase ending_ rather than _end of life starting_.
```suggestion
```

Author: Removed.

### End-of-life

> Six months after entering the *security fixes only*, the minor release goes end-of-life. No more security patches will be shipped for that minor release.

Reviewer 2: ```suggestion
Six months after entering the *security fixes only* phase, the minor release goes *end-of-life*. No more security patches will be shipped for that minor release.
```

Author: Applied suggestion

### The last minor release in a major release line

Shortly before tagging a new major release, one last minor release will be tagged for the major release line in active development.

The last minor release in a major release line will have an extended support phases in line with our major release policy:

> - the last minor stays in the *bug and security fixes* phase for one year

Reviewer 2: ```suggestion
- the last minor release stays in the *bug and security fixes* phase for one year
```

Author: Suggestion applied.

> The last minor release in a major release line will have an extended support phases in line with our major release policy:
> - the last minor stays in the *bug and security fixes* phase for one year
> - then it enters the *security fixes only* phase for another year
> - then the minor release goes end of life along with its corresponding major release line.

Reviewer 2: I'm on the fence about outlining this here, vs just changing it to this:
```suggestion
The last minor release in a major release line will have an extended support phases in line with our [major release policy](major_release_policy).
```

Author: Applied the suggestion in the interest of keeping the doc short and to the point.

> ## Security patch windows

Reviewer 2: Please add a link to this from the `issues_and_bugs/#reporting-security-issues` section.

Author: A link has been added.

Reviewer 2: No it hasn't? There's no changes to the issues_and_bugs page at all?

Author: Oups ... misread this one. Should be there now.

> Silverstripe CMS aims to deliver security patches at pre-defined intervals to minimise upgrades woes for Silverstripe CMS projects.

Reviewer 2: ```suggestion
Silverstripe CMS aims to deliver security patches at pre-defined intervals to minimise upgrade woes for Silverstripe CMS projects.
```

Author: Went with Steve's "plan upgrade" suggestion instead as it uses simpler language.

When a vulnerability is [confidentially reported](/contributing/issues_and_bugs/#reporting-security-issues) to the Silverstripe CMS development team and we are confident that the vulnerability is not actively being exploited in the wild, we will wait until the next available security patch window to disclose it and release a patch.

Security patch windows are scheduled to line up with our expected minor release. Our security patch windows are in:
- January
- April
- July
- October.

> Any security fixes that are shipped at launch with a follow up minor release are also backported to minor currently in the minor release in the "bug and security fixes".

Reviewer 2: ```suggestion
Any security fixes that are shipped at the same time as a new minor release will also be patched and tagged on the minor release that is entering the "bug and security fixes" phase.
```

Author: Suggestion applied.

## Patch releases

> Silverstripe CMS tags patch releases of individual modules whenever relevant bug fixes are merged into the codebase.

Reviewer 2: ```suggestion
Silverstripe CMS tags patch releases of individual modules whenever relevant bug fixes or other very low risk changes are merged into the codebase.
```
e.g. translations have been tagged as patch releases a few times now.

Author: Mix this with Steve's suggestion.

---
title: Minor Release Policy
summary: Outline the minor release lifecycle and support commitment for Silverstripe CMS
icon: code-branch
---

# Minor Release Policy

This policy complements the [Silverstripe CMS Major release policy](major_release_policy) by providing guidance on when new minor releases of Silverstripe CMS are shipped.

Note that the release cadence and pre-release time frame are indicative only. We may alter those in response to external constraints.

> Our minor release policy is more flexible because minor release upgrades are more straight forward to manage than major release upgrades. 

Reviewer 3: ```suggestion
Our minor release policy is more flexible because minor release upgrades are more straightforward to manage than major release upgrades. 
```

Author: Applied suggestion

### Stable release

New minor releases of Silverstripe CMS are tagged around April and October each year.

## Minor release support timeline

The minor release support timeline follows similar phases to those of a major release, but are more condensed.

### Bug and security fixes

A Silverstripe CMS minor release line enters the *bug and security fixes* phase once it is tagged *stable*.

A minor release in the *bug and security fixes* phase receives:
- bugfixes that do not change existing APIs
- security patches for vulnerabilities at all impact levels.

It does **not** receive:
- new features
- new APIs.

A minor release line stays in the *bug and security fixes* phase until a follow up minor release is tagged.

### Security fixes only

At the end of the *bug and security fixes* phase, a Silverstripe CMS minor release line transitions to the *security fixes only* phase.

> In the *security fixes only* phase, a Silverstripe CMS minor release line only receives patches for high impact security issues defined has having a CVSS of 7.0 or above.

Reviewer 3: ```suggestion
In the *security fixes only* phase, a Silverstripe CMS minor release line only receives patches for high impact security issues defined as having a CVSS score of 7.0 or above.
```

Author: Applied suggestion

### Stable release

New minor releases of Silverstripe CMS are tagged around April and October each year.

## Minor release support timeline

The minor release support timeline follows similar phases to those of a major release, but are more condensed.

### Bug and security fixes

A Silverstripe CMS minor release line enters the *bug and security fixes* phase once it is tagged *stable*.

A minor release in the *bug and security fixes* phase receives:
- bugfixes that do not change existing APIs
- security patches for vulnerabilities at all impact levels.

It does **not** receive:
- new feature
- new APIs.

A minor release line stays in the *bug and security fixes* phase until a follow up minor release is tagged.

### Security fixes only

At the end of the *bug and security fixes* phase, a Silverstripe CMS minor release line transitions to the *security fixes only* phase.

In the *security fixes only* phase, a Silverstripe CMS minor release line only receives patches for high impact security issues defined has having a CVSS of 7.0 or above.

> 

Reviewer 2: ```suggestion
```

Author: Removed.
